### DISCUSSION TITLE
Proposal: readonly local variables + Explicit 'mutable' specifier to help with code analysis + Related compiler settings

### DISCUSSION THREAD

**Discussion by Lehonti** [dotnet/csharplang#3746] [Language Ideas]
Currently, fields can be assigned and re-assigned by default, unless an explicit `readonly` modifier is used. `readonly` fields exist to help the programmer ensure certain invariants during the lifetime of an object.

My first proposal is to provide a similar functionality for local variables. They would have to be assigned a value during declaration and this would ensure certain invariants during a function call. Something like this:

``` C#
void DoSomething (ExecutionContext context) {
    readonly int number = _numberFactory.Create(context); // OK
    number = 4; // Error: 'number' can't be assigned a value after declaration since it is read-only.
}
```

My second proposal is to create a new `mutable` modifier for fields and variables, which would allow us to explicitly specify the semantics that fields now have by default (when the `readonly` modifier is not present). This would help code analysis. Someone could configure their compiler such that fields and variables without any specifier are treated as `readonly` by default, or at least give info/warnings/errors when a non-`mutable` field or variable is re-assigned to, just like we can currently configure the compiler to allow for non-nullable reference types to exist.

**Edit**: It would also be a good idea to have readonly parameters, as proposed [here](https://github.com/dotnet/csharplang/issues/188). Thanks, @svick

**Comment by svick** [comment]
> My first proposal is to provide a similar functionality for local variables. 

That sounds like a duplicate of https://github.com/dotnet/csharplang/issues/188.

> My second proposal is to create a new `mutable` modifier for fields and variables

At least for fields, that sounds like something that could be achieved using an attribute.

**Comment by Lehonti** [comment]
Thanks @svick for noting that my proposal includes something that was already suggested. Still, I think it makes the most sense in combination with the `mutable` part. Also, if attributes could be applied to local variables, I could probably buy the solution you are proposing. Sadly, this is not the case as far as I know.

**Comment by svick** [comment]
> if attributes could be applied to local variables

There's a proposal for that too: https://github.com/dotnet/csharplang/issues/2478. Though it hasn't seen any movement in quite some time.

**Comment by Richiban** [comment]
Wow. #188. That issue was championed 3 ½ years ago!

I know about the prioritisation process (roughly) for features but, damn, I thought this one would have been chosen for development by now.
